### ISSUE TITLE
Error installing tidyverse on macOS 10.10.5/R 3.4.1 

### ISSUE THREAD

**Issue by klmr** [tidyverse/tidyverse#70]
Attempting to install ‹tidyverse› on a blank-slate installation fails due to what seems like a circular dependency of the packages that therefore cannot be unloaded in the last step of the installation (that test whether the package can be loaded).

Here’s an MWE:

```r
dir.create('test-library')
.libPaths('test-library')
install.packages('tidyverse')
```

This yields:

```
Installing package into ‘/Users/rudolph/tmp/test-library’
(as ‘lib’ is unspecified)
also installing the dependencies ‘colorspace’, ‘mnormt’, ‘bindr’, ‘RColorBrewer’, ‘dichromat’, ‘munsell’, ‘labeling’, ‘rematch’, ‘plyr’, ‘psych’, ‘reshape2’, ‘assertthat’, ‘bindrcpp’, ‘glue’, ‘pkgconfig’, ‘rlang’, ‘R6’, ‘Rcpp’, ‘BH’, ‘plogr’, ‘digest’, ‘gtable’, ‘scales’,
 ‘lazyeval’, ‘mime’, ‘curl’, ‘openssl’, ‘cellranger’, ‘stringi’, ‘selectr’, ‘broom’, ‘dplyr’, ‘forcats’, ‘ggplot2’, ‘haven’, ‘httr’, ‘hms’, ‘jsonlite’, ‘lubridate’, ‘magrittr’, ‘modelr’, ‘purrr’, ‘readr’, ‘readxl’, ‘stringr’, ‘tibble’, ‘rvest’, ‘tidyr’, ‘xml2’
[…]
trying URL 'https://cran.ma.imperial.ac.uk/src/contrib/tidyverse_1.1.1.tar.gz'
Content type 'application/x-gzip' length 23452 bytes (22 KB)
==================================================
downloaded 22 KB
* installing *source* package ‘tidyverse’ ...
** package ‘tidyverse’ successfully unpacked and MD5 sums checked
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error: package or namespace load failed for ‘tidyverse’:
 .onAttach failed in attachNamespace() for 'tidyverse', details:
  call: value[[3L]](cond)
  error: Package ‘tidyr’ version 0.6.3 cannot be unloaded:
 Error in unloadNamespace(package) : namespace ‘tidyr’ is imported by ‘broom’, ‘tidyverse’ so cannot be unloaded

Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/Users/rudolph/tmp/test-library/tidyverse’

The downloaded source packages are in
        ‘/private/var/folders/9t/zz9zg0v914z_64yc4v2qtcp40000gn/T/RtmpDSaA9l/downloaded_packages’
Warning message:
In install.packages("tidyverse") :
  installation of package ‘tidyverse’ had non-zero exit status
```

Rerunning the installation yields the same error every time. It sees as if all the constituent packages are actually successfully installed — just not ‹tidyverse› itself.

```
> sessionInfo()
R version 3.4.1 (2017-06-30)
Platform: x86_64-apple-darwin14.5.0 (64-bit)
Running under: OS X Yosemite 10.10.5

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
[1] compiler_3.4.1 tools_3.4.1
```

**Comment by jbromley** [comment]
On Ubuntu 16.04.3 with R 3.4.1 I get something similar when trying klmr's test case, except that the package that can't be unloaded is 'tibble'. Here is the relevant portion of the output:

```
* installing *source* package ‘tidyverse’ ...
** package ‘tidyverse’ successfully unpacked and MD5 sums checked
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error: package or namespace load failed for ‘tidyverse’:
 .onAttach failed in attachNamespace() for 'tidyverse', details:
  call: value[[3L]](cond)
  error: Package ‘tibble’ version 1.3.4 cannot be unloaded:
 Error in unloadNamespace(package) : namespace ‘tibble’ is imported by ‘ggplot2’, ‘tidyr’, ‘modelr’, ‘readr’, ‘dplyr’, ‘haven’ so cannot be unloaded

Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/home/jay/Development/packrat_test/test-library/tidyverse’

The downloaded source packages are in
        ‘/tmp/RtmpJ7uB2e/downloaded_packages’
Warning message:
In install.packages("tidyverse") :
  installation of package ‘tidyverse’ had non-zero exit status
```
My session information is:

```
* installing *source* package ‘tidyverse’ ...
** package ‘tidyverse’ successfully unpacked and MD5 sums checked
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error: package or namespace load failed for ‘tidyverse’:
 .onAttach failed in attachNamespace() for 'tidyverse', details:
  call: value[[3L]](cond)
  error: Package ‘tibble’ version 1.3.4 cannot be unloaded:
 Error in unloadNamespace(package) : namespace ‘tibble’ is imported by ‘ggplot2’, ‘tidyr’, ‘modelr’, ‘readr’, ‘dplyr’, ‘haven’ so cannot be unloaded

Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/home/jay/Development/packrat_test/test-library/tidyverse’

The downloaded source packages are in
        ‘/tmp/RtmpJ7uB2e/downloaded_packages’
Warning message:
In install.packages("tidyverse") :
  installation of package ‘tidyverse’ had non-zero exit status
```
I have this same problem inside a packrat environment as well.

**Comment by hadley** [comment]
Just to be sure, you're running this in a clean R session with no packages attached by your `.Rprofile`?

**Comment by hadley** [comment]
I can't reproduce on mac with binary packages; now trying source packages.

**Comment by hadley** [comment]
Hmmm, I see it too (although with slightly different packages):

```
Error: package or namespace load failed for ‘tidyverse’:
 .onAttach failed in attachNamespace() for 'tidyverse', details:
  call: value[[3L]](cond)
  error: Package ‘ggplot2’ version 2.2.1 cannot be unloaded:
 Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘tidyverse’ so cannot be unloaded
```

@jimhester any idea why this is happening?

**Comment by klmr** [comment]
Yep, this is in a clean session with an empty `.Rprofile`. That said, I cannot reproduce this on a new mac with macOS 10.12.6.

**Comment by hadley** [comment]
If you have the time/patience would you mind rerunning a few times? I think it might be stochastic. This is almost certainly a bug in install.pacakges() but it would be interesting to know more

**Comment by klmr** [comment]
I can do that, but what am I looking for as a diagnostic readout?

At the moment I’m running the installation in a directory with the following `.Rprofile` in it:

```r
options(repos = c(CRAN = 'https://cloud.r-project.org/'))
dir.create('test-library', showWarnings = FALSE)
.libPaths('test-library')
```

So far, it consistently fails but I’ll keep running it (deleting the `test-library` directory in between runs).

**Comment by hadley** [comment]
I'm just interested if it always fails on the same package - it looks like you, me and @jbromley all had different failures.

**Comment by klmr** [comment]
Oh. Yes, it always fails on the same packages for me. In fact, the output is identical every time (I record the standard output and stderr and `diff` them).

**Comment by hadley** [comment]
@jimhester notes that the error comes from https://github.com/wch/r-source/blob/d1f01dc790201e2e40c97e8da45c51c495874ffe/src/library/base/R/library.R#L283.

The source of the problem seems to be the way in which I'm attaching extra packages in `.onAttach()`.

**Comment by hadley** [comment]
Do you have `R_LIBS` set in `.Renviron` or similar?

**Comment by hadley** [comment]
I think this is a bug in `install.packages()`/`library()`. The root error comes from `library()` - if you attach a package that is already loaded, `library()` compares the package version between the loaded version and the on disk version, and attempts to [unload and reload](https://github.com/wch/r-source/blob/d1f01dc790201e2e40c97e8da45c51c495874ffe/src/library/base/R/library.R#L277-L293).  `library()` is called by [`.test_load_package()`](https://github.com/wch/r-source/blob/89ec1150299f7be62b839d5d5eb46bd9a63653bd/src/library/tools/R/admin.R#L918), which is called by [`install.packages`](https://github.com/wch/r-source/blob/17ea295025dcd71a5834a787c51380a358e7274c/src/library/tools/R/install.R#L1213) as part of a check to ensure the package has been installed somewhere. `.test_load_package()` is run in another R session but not with `--vanilla`.

My hypothesis is that somewhere in the bowels of `library()` a function doesn't pass `lib.loc` along, so it ends up using the version of the package you currently have installed in one place, and the version you're trying to install in another. This is hypothesis is supported by two fixes that work for me:

* Remove `R_LIBS` from my `~/.Renviron`
* Install ggplot2 from CRAN

I think this explains why we each see failures with slightly different packages: I have a dev version of ggplot2 and I'm guessing that @klmr has dev tidyr and @jbromley has dev tibble.

**Comment by hadley** [comment]
I can reliably reproduce with:

```R
install_github("tidyverse/ggplot2")
packageVersion("ggplot2")

tmp <- tempfile()
dir.create(tmp)
.libPaths(tmp)
install.packages("tidyverse", type = "source")
```

(But note this takes a loooooong time)

**Comment by klmr** [comment]
> Do you have `R_LIBS` set in `.Renviron` or similar?

No — I just verified this and it’s set neither in `$R_HOME` nor in `$HOME` (nor the test directory). To be on the safe side, I’ve deleted my `$HOME/.Renviron` file (it was obsolete anyway). Also, to verify, inside the test environment, my `.libPaths()` shows only this:

```
[1] "/Users/rudolph/tmp/tinytidy/test-library"
[2] "/usr/local/Cellar/r/3.4.1_1/R.framework/Versions/3.4/Resources/library"
```

… and I have almost no libraries installed in the Framework path beyond the default ones.

And unfortunately for your hypothesis I haven’t got dev tidyr installed *anywhere* (it’s 0.7.1, from CRAN; same for broom, which my error message also mentions).

**Comment by hadley** [comment]
Bummer :( I have no idea what's going wrong for you then, except I'm still fairly certain its `install.packages()` fault.

**Comment by Titan100** [comment]
Hi @hadley and all,
I am having the same issue after the recommended installation `tmp <- tempfile()
dir.create(tmp)
.libPaths(tmp)
install.packages("tidyverse", type = "source")`

It looks like it is common problem. Any suggestions yet?

`> sessionInfo()
R version 3.3.3 (2017-03-06)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS  10.13.3

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] tools_3.3.3 tcltk_3.3.3`

**Comment by dnmabra12** [comment]
I was getting the same error, and then specified the library location when loading:

library(stringr,libloc = "LibraryLocation")

Error went away and stringr worked again. Not sure why. Maybe @hadley has some insight on that.

**Comment by kdmurray91** [comment]
In case anyone else is looking for a workaround, `devtools::install_cran` was able to install `tidyverse` even when `install.packages` was refusing to due to this issue.

**Comment by dpastoor** [comment]
This is still presenting, I have tried pointing to multiple points in CRAN time (through using the microsoft MRAN snapshots). I went back ~ 1 month at 1 week intervals and in all instances it failed.

It seems to fail consistently on ggplot2:

```
Error: package or namespace load failed for ‘tidyverse’:
 .onAttach failed in attachNamespace() for 'tidyverse', details:
  call: value[[3L]](cond)\n  error: Package ‘ggplot2’ version 3.1.0 cannot be unloaded:
 Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘tidyverse’ so cannot be unloaded
Error: loading failed
Execution halted
```

I have tried this through install.packages("tidyverse", type = "source") as well as pulling the tarball and running direclty through R CMD INSTALL after all other dependencies are installed with the same outcome.

The installation succeeds when installing directly through binary.

**Comment by hadley** [comment]
@dpastoor it sounds like you have ggplot2 loaded. Try installing in a fresh R session.

**Comment by maurolepore** [comment]
Is this a bad idea? Or is anyone developing this already?

In a meta-package inspired by the __tidyverse__, I'm drafting an install-helper (think `tidyverse::install_me()`) which goal is to avoid installing via DESCRIPTION > Imports and instead use something like the `please_install()` function that @hadley wrote to set-up his tidy-tools workshop.

Benefits of `install_me()` :

* installs directly from CRAN avoiding the `Error in unloadNamespace(package) : namespace ‘PKG’ is imported by ‘tidyverse’ so cannot be unloaded`.
* Catches the errors and throws more informative messages.
* Captures the output of  `intall.packages()` and shows progress, errors, warnings and messages as __testthat__ does during testing. This way users can focus on the important stuff.

**Comment by hadley** [comment]
@maurolepore we are working on this already; it's a lot of work to do correctly.

**Comment by dpastoor** [comment]
@hadley completely fresh session(s). I even worried that maybe the daily version of Rstudio was loading stuff so tried from both the terminal via `install.packages`, and finally, this even presents when installing via R CMD INSTALL on the command line after manually downloading and pointing to the tidyverse tarball

**Comment by maurolepore** [comment]
@hadley, great! It that work public? I'd love to follow that.

**Comment by hadley** [comment]
@dpastoor are you sure you're not loading packages in your `.Rprofile`? If so, try deleting your entire library.

@maurolepore yes, but it's not suitable for public consumption yet so we're not advertising it, and you'll need to find it yourself.

**Comment by dpastoor** [comment]
@hadley I think I'm tracking it down. It looks like it is an artifact somehow of in the tmp directory where the installation occurs, if tidyverse deps are present is causing this corruption

For example, given the following command:

```
R CMD INSTALL --build --install-tests --no-multiarch --with-keep.source --library=/var/folders/km/ljmnf0bx73l9m4zdygvb6q_80000gn/T/ /Users/devinp/go/src/github.com/dpastoor/rpackagemanager/integration_tests/testsets/pkgcache2/CRAN/src/tidyverse_1.2.1.tar.gz
```

if in `/var/folders/km/ljmnf0bx73l9m4zdygvb6q_80000gn/T/` there are still folders such as `tidyr` folder that contains an installed version of tidyr, 

```
* installing *source* package ‘tidyverse’ ...
** package ‘tidyverse’ successfully unpacked and MD5 sums checked
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded\nERROR: loading failed
* removing ‘/private/var/folders/km/ljmnf0bx73l9m4zdygvb6q_80000gn/T/tidyverse’" Error: package or namespace load failed for ‘tidyverse’:
 .onAttach failed in attachNamespace() for 'tidyverse', details:
  call: value[[3L]](cond)\n  error: Package ‘tidyr’ version 0.8.2 cannot be unloaded:
 Error in unloadNamespace(package) : namespace ‘tidyr’ is imported by ‘broom’, ‘tidyverse’ so cannot be unloaded
Error: loading failed
Execution halted
```

By clearing out the tmpdir and/or creating a nested subfolder that is empty, the error resolves, though I'm not entirely clear what is causing the conflict in the first place :-/

**Comment by alexiswl** [comment]
I was also able to reproduce this error on CentOS7 and was a bug inside my .Rprofile. To reduce issues being unable to write to shared locations I had placed the following in my .Rprofile such that packages would be installed only into my home directory without R trying to update packages that existed in the shared parths.

`assign(".lib.loc", Sys.getenv("R_LIBS_USER"), envir = environment(.libPaths))`

However, when using packrat, installing packages such as tidyverse which restart the rsession during installation then cause issues when packrat is installing the package into the project directory.

In my case wrapping the assignment in the following if statement solved the issue.
```
 if "packrat" %nin% loadedNamespaces() {
  assign(".lib.loc", Sys.getenv("R_LIBS_USER"), envir = environment(.libPaths))
}
```

**Comment by bovender** [comment]
Had the very same error on my machine after it upgraded R from 3.5.2 to 3.5.3. This is with a project that uses packrat.

In my case it was resolved by removing a call to `.libPaths()` from my `.Rprofile` file.

    # ~/.Rprofile
    # .libPaths(c("/home/me/.R", .libPaths()))

(Might just as well have deleted the file.)

**Comment by hadley** [comment]
Will probably be fixed by #171

**Comment by hadley** [comment]
Confirmed that this is fixed with:

```R
tmp <- tempfile()
dir.create(tmp)
.libPaths(tmp)

install.packges("xml2")
remotes::install_github("tidyverse/tidyverse", type = "source")
```

Installation failed when using the CRAN version:

```R
tmp <- tempfile()
dir.create(tmp)
.libPaths(tmp)

install.packges("tidyverse", type = "source")
```
